contrib/mesos/pkg/scheduler/podtask/procurement.go

```diff
@@ -0,0 +1,145 @@
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package podtask
+
+import (
+	"strings"
+
+	log "github.com/golang/glog"
+	mesos "github.com/mesos/mesos-go/mesosproto"
+	mresource "k8s.io/kubernetes/contrib/mesos/pkg/scheduler/resource"
+)
+
+// NewDefaultProcurement returns the default procurement strategy that combines validation,
+// node, pod-resource, and ports procurement. c and m are resource quantities written into
+// k8s api.Pod.Spec's that don't declare resources (all containers in k8s-mesos require cpu
+// and memory limits).
+func NewDefaultProcurement(c mresource.CPUShares, m mresource.MegaBytes) Procurement {
+	requireSome := &RequireSomePodResources{
+		defaultContainerCPULimit: c,
+		defaultContainerMemLimit: m,
+	}
+	return AllOrNothingProcurement([]Procurement{
+		ValidateProcurement,
+		NodeProcurement,
+		requireSome.Procure,
+		PodResourcesProcurement,
+		PortsProcurement,
+	}).Procure
+}
+
+// Procurement funcs allocate resources for a task from an offer.
+// Both the task and/or offer may be modified.
+type Procurement func(*T, *mesos.Offer) error
+
+type AllOrNothingProcurement []Procurement
+
+func (a AllOrNothingProcurement) Procure(t *T, details *mesos.Offer) error {
+	for _, p := range a {
+		if err := p(t, details); err != nil {
+			t.Reset()
+			return err
+		}
+	}
+	return nil
+}
+
+// ValidateProcurement checks that the offered resources are kosher, and if not panics.
+// If things check out ok, t.Spec is cleared and nil is returned.
+func ValidateProcurement(t *T, details *mesos.Offer) error {
+	if details == nil {
+		//programming error
+		panic("offer details are nil")
+	}
+	t.Spec = Spec{}
+	return nil
+}
+
+// NodeProcurement updates t.Spec in preparation for the task to be launched on the
+// slave associated with the offer.
+func NodeProcurement(t *T, details *mesos.Offer) error {
+	t.Spec.SlaveID = details.GetSlaveId().GetValue()
+	t.Spec.AssignedSlave = details.GetHostname()
+
+	// hostname needs of the executor needs to match that of the offer, otherwise
+	// the kubelet node status checker/updater is very unhappy
+	const HOSTNAME_OVERRIDE_FLAG = "--hostname-override="
+	hostname := details.GetHostname() // required field, non-empty
+	hostnameOverride := HOSTNAME_OVERRIDE_FLAG + hostname
+
+	argv := t.executor.Command.Arguments
+	overwrite := false
+	for i, arg := range argv {
+		if strings.HasPrefix(arg, HOSTNAME_OVERRIDE_FLAG) {
+			overwrite = true
+			argv[i] = hostnameOverride
+			break
+		}
+	}
+	if !overwrite {
+		t.executor.Command.Arguments = append(argv, hostnameOverride)
```

Comment:
Yes, but will kube-proxy? In master they've updated kube-proxy to generate
events to the apiserver, and kube-proxy needs this value.

On Thu, Sep 3, 2015 at 4:56 AM, Dr. Stefan Schimanski <
notifications@github.com> wrote:

> In contrib/mesos/pkg/scheduler/podtask/procurement.go
> <https://github.com/kubernetes/kubernetes/pull/13421#discussion_r38625087>
> :
>
> > +	// the kubelet node status checker/updater is very unhappy
> > +	const HOSTNAME_OVERRIDE_FLAG = "--hostname-override="
> > +	hostname := details.GetHostname() // required field, non-empty
> > +	hostnameOverride := HOSTNAME_OVERRIDE_FLAG + hostname
> > +
> > +	argv := t.executor.Command.Arguments
> > +	overwrite := false
> > +	for i, arg := range argv {
> > +		if strings.HasPrefix(arg, HOSTNAME_OVERRIDE_FLAG) {
> > +			overwrite = true
> > +			argv[i] = hostnameOverride
> > +			break
> > +		}
> > +	}
> > +	if !overwrite {
> > +		t.executor.Command.Arguments = append(argv, hostnameOverride)
>
> This will go away with my "decouple executor + kubelet" patch. Then the
> kubelet will know from the SlaveInfo who it is.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/13421/files#r38625087>.
>